docs/specs/user-profile/design.md

```diff
@@ -147,49 +147,98 @@ In addition to standard attributes, the developer can define their own custom at
 
 ### Defining custom attributes
 
-The custom attributes are defined as a single JSON schema written against a subset of JSON schema 2019-09.
+The custom attributes are defined with a configuration.
 
-Here is an example of the schema of the custom attributes.
+Here is an example of the configuration.
 
-```yaml
+```
 user_profile:
   custom_attributes:
-    schema:
-      properties:
-        stripe_customer_id:
-          type: string
-        hobby:
-          type: string
+    attributes:
+    - pointer: /x_phone_number
+      type: phone_number
+    - pointer: /x_email
+      type: email
 ```
 
-All changes made to custom attributes must be valid against the schema.
+Each custom attribute MUST have an unique pointer and of one of the defined types.
+
+The pointer of custom attribute MUST have exactly ONE level, as in `/this_is_one_level`.
+The pointer MUST also be non-empty, so `/` is not a valid pointer.
+
+The pointer MUST NOT conflict with the pointer of any standard attributes,
+so the developer CANNOT define a custom attribute with pointer `/email`.
+
+NO data migration is ever done when the configuration is changed.
+This means if the developer change the pointer, existing data WILL BE lost.
```

Comment:
How about changing the type and other validation config of an existing field?
We will do the validation during read/write or write only?

Reply:
It is possible to change the type if the underlying type is still the same. We validate the whole custom attributes object so if the underlying type has been changed, the validation will fail. 

For example, let say the developer initially define `a: integer` and `b: string`. Some users have `a` being set. The developer then change to `a: string`. The underlying type has been changed. So when a user attempts to change `b`, `a` will fail the validation.

An example of changing type is safe is type widening, like change from `a: enum` to `a: string`. It is because the old type is a subset of the new type so validation must succeed.

Reply:
It is `portal_ui`.

Reply:
> So when a user attempts to change b, a will fail the validation.

That means user will need to correct `a` before changing `b`. Right? 

There is a rare case that if it is type narrowing / changing (string to enum, string to number), the initial value will not be able to show in the ui and the ui field should be in empty state. User will need to fill in value in order to save the form. 

I think it is acceptable, given that type changing is not a common operation. Maybe add warning when developer changes the custom attributes config in the portal?

> It is portal_ui.

Please update the table `Exhaustive list of access control combination`:)

Reply:
The editing UI for standard attribute is editing one standard attribute at a time. Custom attributes should follow this design. When the end-user wants to change `b`, they can only see the control for changing `b`. There is no way for them to be aware that `a` needs correction. Anyway, I think it should not be the responsibility of the end-user to correct `a` themselves.

Actually there is another case that can change the type of a custom attribute. It is by deleting the custom attribute definition first, followed by defining a new one with an incompatible type and the same name :(

Reply:
> It is by deleting the custom attribute definition first, followed by defining a new one with an incompatible type and the same name.

yes, I thought of it too :(

> So when a user attempts to change `b`, `a` will fail the validation.

> The editing UI for standard attribute is editing one standard attribute at a time.

Then how can we resolve this case? Change the api to allow user to update one attribute and only that attribute will be validated? So `a` will keep the initial value when user change `b`. Or any other ideas?

Reply:
Discussed with Fung, I am going to follow [the approach of Amazon Cognito](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html). Custom attributes, once defined, cannot be removed or changed. With this limitation introduced, this problem will be eliminated. I am going to update the spec to specify how this approach works with our own design.